routesrv/polling.go

```diff
@@ -0,0 +1,111 @@
+package routesrv
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	ot "github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/routing"
+	"github.com/zalando/skipper/tracing"
+)
+
+const (
+	LogPollingStarted       = "starting polling"
+	LogPollingStopped       = "polling stopped"
+	LogRoutesFetchingFailed = "failed to fetch routes"
+	LogRoutesEmpty          = "received empty routes; ignoring"
+	LogRoutesInitialized    = "routes initialized"
+	LogRoutesUpdated        = "routes updated"
+)
+
+var (
+	pollingStarted = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "routesrv",
+		Name:      "polling_started_timestamp",
+		Help:      "UNIX time when the routes polling has started",
+	})
+	routesInitialized = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "routesrv",
+		Name:      "routes_initialized_timestamp",
+		Help:      "UNIX time when the first routes were received and stored",
+	})
+	routesUpdated = promauto.NewGauge(prometheus.GaugeOpts{
+		Namespace: "routesrv",
+		Name:      "routes_updated_timestamp",
+		Help:      "UNIX time of the last routes update (initial load counts as well)",
+	})
+)
+
+type poller struct {
+	client  routing.DataClient
+	b       *eskipBytes
+	timeout time.Duration
+	quit    chan struct{}
+
+	tracer ot.Tracer
+}
+
+func (p *poller) poll(wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	var (
+		routesCount, routesBytes int
+		initialized              bool
+		msg                      string
+	)
+
+	log.WithField("timeout", p.timeout).Info(LogPollingStarted)
+	pollingStarted.SetToCurrentTime()
+	for {
+		span := tracing.CreateSpan("poll_routes", context.TODO(), p.tracer)
+
+		routes, err := p.client.LoadAll()
+		routesCount = len(routes)
+
+		switch {
+		case err != nil:
+			log.WithError(err).Error(LogRoutesFetchingFailed)
+
+			span.SetTag("error", true)
+			span.LogKV(
+				"event", "error",
+				"message", fmt.Sprintf("%s: %s", LogRoutesFetchingFailed, err),
+			)
+		case routesCount == 0:
+			log.Error(LogRoutesEmpty)
+
+			span.SetTag("error", true)
+			span.LogKV(
+				"event", "error",
+				"message", msg,
+			)
+		case routesCount > 0:
+			routesBytes, initialized = p.b.formatAndSet(routes)
+			logger := log.WithFields(log.Fields{"count": routesCount, "bytes": routesBytes})
+			if initialized {
+				logger.Info(LogRoutesInitialized)
+				span.SetTag("routes.initialized", true)
```

Comment:
Because only when we have some routes (`initialized`) our routes are considered valid, thus we are able to serve successfully to Skipper. Otherwise we can serve, however potentially with no use for Skipper. Specifically for Kubernetes data client no routes is not considered a valid state/update and is discarded as far as I understand the current implementation.

Reply:
So we want to know / measure time from start to being setup to serve valid routes.